src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,507 @@
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe, ucc_rank_t rank,
+    size_t put_window_size)
 {
+    int i;
+
+    pipe->avail_buffs = pipe->num_buffers;
+    pipe->src_rank = pipe->dst_rank = rank;
+    pipe->get_idx = pipe->red_idx = 0;
+    pipe->done_get = pipe->done_red = 0;
+    pipe->done_put = pipe->posted_put = 0;
+    pipe->count_issued = pipe->count_received = 0;
+    pipe->count_reduced = pipe->count_serviced = 0;
+    pipe->my_count = pipe->my_offset = 0;
+
+    ucc_tl_ucp_allreduce_sliding_window_reset_buf(&pipe->accbuf);
+    for (i = 0; i < pipe->num_buffers; i++) {
+        ucc_tl_ucp_allreduce_sliding_window_reset_buf(&pipe->getbuf[i]);
+    }
+
+    for (i = 0; i < put_window_size; i++) {
+        pipe->put_requests[i] = NULL;
+    }
 }
 
 ucc_status_t
-ucc_tl_ucp_allreduce_sliding_window_start(ucc_coll_task_t __attribute__((unused)) *coll_task) //NOLINT
+ucc_tl_ucp_allreduce_sliding_window_start(ucc_coll_task_t *coll_task)
 {
-    return UCC_OK;
+    ucc_status_t       status = UCC_OK;
+    ucc_tl_ucp_task_t *task   = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team   = TASK_TEAM(task);
+    ucc_rank_t         rank   = UCC_TL_TEAM_RANK(team);
+    uint32_t           count_total = coll_task->bargs.args.dst.info.count;
+    ucc_rank_t         size        = coll_task->team->params.size;
+    ucc_datatype_t     dtype       = TASK_ARGS(task).dst.info.datatype;
+    size_t             dt_size     = ucc_dt_size(dtype);
+    int                put_window_size = UCC_TL_UCP_TEAM_LIB(team)
+                                 ->cfg.allreduce_sliding_window_put_window_size;
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe =
+        task->allreduce_sliding_window.pipe;
+    ucc_tl_ucp_allreduce_sw_host_allgather_t *allgather_data =
+        task->allreduce_sliding_window.allgather_data;
+    size_t allgather_size = sizeof(ucc_tl_ucp_allreduce_sw_host_allgather_t);
+
+    ucc_base_coll_args_t bargs    = {
+        .args = {
+            .coll_type = UCC_COLL_TYPE_ALLGATHER,
+            .mask      = 0,
+            .src.info = {.buffer   = allgather_data,
+                         .count    = allgather_size,
+                         .datatype = UCC_DT_UINT8,
+                         .mem_type = UCC_MEMORY_TYPE_HOST},
+            .dst.info = {.buffer   = PTR_OFFSET(allgather_data, allgather_size),
+                         .count    = allgather_size * size,
+                         .datatype = UCC_DT_UINT8,
+                         .mem_type = UCC_MEMORY_TYPE_HOST}
+        }
+    };
+
+    UCC_CHECK_GOTO(ucc_tl_ucp_allgather_knomial_init(&bargs,
+                    &team->super.super,
+                    &task->allreduce_sliding_window.allgather_task),
+        out, status);
+
+    task->allreduce_sliding_window.allgather_task->executor =
+                                        task->allreduce_sliding_window.executor;
+
+    UCC_CHECK_GOTO(ucc_tl_ucp_allgather_knomial_start(
```

Comment:
same remark as for barrier, IIUC we have
allgather -> allreduce -> barrier
In this case this can be scheduled during init and managed automatically by the core.

Reply:
Thank you, same thing as the barrier--https://github.com/openucx/ucc/pull/958#discussion_r1602109352
Please let me know if that's ok.